nfpm.native_libs: deb search_for_sonames script to find pkg deps

cognifloyd opened the pull request:
## PR Series Overview

This is the third in a series of PRs that introduces a new backend: `pants.backend.npm.native_libs`
Initially, the backend will be available as: `pants.backend.experimental.nfpm.native_libs`

I proposed this new backend (originally named `bindeps`) in discussion #22396.

This backend will inspect ELF bin/lib files (like `lib*.so`) in packaged contents (for this PR series, only in `pex_binary` targets) to identify package dependency metadata and inject that metadata on the relevant `nfpm_deb_package` or `nfpm_rpm_package` targets. Effectively, it will provide an approximation of these native packager features:
- `rpm`: `rpmdeps` + `elfdeps`
- `deb`: `dh_shlibdeps` + `dpkg-shlibdeps` (These substitute `${shlibs:Depends}` in debian control files have)

### Goal: Host-agnostic package builds

This pants backend is designed to be host-agnostic, like [nFPM](https://nfpm.goreleaser.com/).

Native packaging tools are often restricted to a single release of a single distro. Unlike native package builders, this new pants backend does not use any of those distro-specific or distro-release-specific utilities or local package databases. This new backend should be able to run (help with building deb and rpm packages) anywhere that pants can run (MacOS, rpm linux distros, deb linux distros, other linux distros, docker, ...).

### Previous PRs in series

- #22873
- #22899

## PR Overview

To achieve the host-agnostic goal, this PR introduces the `nfpm.native_libs.deb` package that contains a `search_for_sonames` script. This script uses a debian HTML API to figure out which package provides a given SONAME. Normally, deb helpers (native deb build tooling) rely on metadata files from installed packages or local databases to figure this out. Using the API allows this backend to be host-agnostic instead.

### Dependencies

The `search_for_sonames` script requires `aiohttp`, `aiohttp-retries`, and `beautifulsoup4`. This script is treated as "internal" pants code for which users cannot change the dependency versions.

I couldn't think of any reasons to deal with the added complexity of yet another subsystem and lockfile. So, I added these deps to `3rdparty/python/requirements.txt`, and then pull those package versions from the pants venv when running the script that needs them.

```
__________________________________________________________________
Lockfile diff: 3rdparty/python/user_reqs.lock [python-default]
__________________________________________________________________
==                      Added dependencies                      ==
‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
  aiohappyeyeballs               2.6.1
  aiohttp                        3.12.15
  aiohttp-retry                  2.9.1
  aiosignal                      1.4.0
  attrs                          25.4.0
  frozenlist                     1.8.0
  multidict                      6.7.0
  propcache                      0.4.1
  yarl                           1.22.0
```

> [!NOTE]
> The `search_for_sonames` script uses `asyncio` (and therefore does not use synchronous `requests`) because it could involve many IO-bound API calls (all of which should be cached using standard pants caching). As the dep is included with pants, I chose to use `aiohttp` because it's smaller than `httpx`. But, `httpx` has a more modern feature set, so if something like HTTP/2 proves to be an important optimization, the script could be refactored to use it instead.

### Rules

The `deb_search_for_sonames` rule:

- create a pex venv that:

    - includes the `search_for_sonames` script
    - includes script dependencies constrained to the locked versions present in the pants venv.

- Run script `pants.backend.nfpm.native_libs.deb.search_for_sonames` in pex venv.
- Parse the `search_for_soname` script's JSON output, returning it in a dataclass that offers better structured data access than the HTML API response.

### `search_for_sonames.py` script

The `search_for_sonames.py` script in `pants.backend.nfpm.native_libs.deb` does the following:

- Issue HTTPS package search API calls which returns results in HTML
  (the HTML-based API design predates JSON and other web API technologies).
- Retry API calls up to 5 times with jitter + exponential backoff
  (because the API is somewhat flaky, often returning 500 when--I assume--it's overloaded).
- Use `beautifulsoup4` to extract the package names from the table in the HTML API response
  (this data structure has been consistent for decades, so script updates/maintenance should be minimal).

### Tests

- `src/python/pants/backend/nfpm/native_libs/deb/search_for_sonames_test.py`:
    - `test_deb_packages_from_html_response`: A unit test that ensures the `beautifulsoup4`-based HTML response parsing is working as expected

- `src/python/pants/backend/nfpm/native_libs/deb/search_for_sonames_integration_test.py`:
    - `test_deb_search_for_sonames`: Test the primary logic (the `deb_search_for_soname()` function) in the script by itself
    - `test_deb_search_for_sonames_rule`: Test the script's `main()` function, and the rule that builds the venv and actually runs the script.

Logically, it would like to split the integration tests into separate files (`rules_integration_test.py` and `search_for_sonames_integration_test.py`), so that testing the script is separate from testing the rule that runs the script. But, I'm sharing the `TEST_CASES` for both tests, and putting the test cases in a separate file from the tests feels like a code-smell as it makes it harder to follow what the tests are actually testing. I'm open to alternative suggestions on how to structure these tests.

cognifloyd committed:
nfpm.native_libs.scripts: add aiohttp requirement

This commit was rebased on main. Before rebase, the lockfile was
regenerated. After rebase, the lockfile regeneration is batched into a
single follow-up commit.

cognifloyd committed:
nfpm.native_libs.scripts: Add deb_search_for_sonames with tests

includes:
- fix flake8 C413 Unnecessary list call around sorted(
- add type annotation for mypy

cognifloyd committed:
nfpm.native_libs: add deb_search_for_sonames rule

The rule runs the deb_search_for_sonames.py script.
It pulls the pex requirements from the pants venv so that the script
runs with the same version of python
and a subset of the dists used to run pants itself.
This way, the deps are only defined once.

cognifloyd committed:
nfpm.native_libs: add integration test for scripts.deb_search_for_sonames rule

This is in the scripts integration test file instead of the one for rules to facilitate sharing TEST_CASES for both tests.

cognifloyd committed:
nfpm.native_libs.scripts: Refactor to preserve API response data

We need to select the relevant package(s) based on the .so file path,
following (a simplified set of) standard LD lookup rules.
This commit updates the deb_search_for_sonames script and rule
so that .so file names are available for that selection process.

cognifloyd committed:
nfpm.native_libs.scripts: Add aiohttp-retry requirement

Also registers script dep on aiohttp-retry

This commit was rebased. Lockfile was regenerated before rebase, but is
batched up with other regenerations after the rebase.

cognifloyd committed:
nfpm.native_libs.scripts: Use aiohttp-retry on flaky API

NOTE: order of client and task group in async with block is very
important. The task group needs to be the innermost block so that it can
appropriately wait for the task group to finish. Putting the client as
the innermost block closes the client before any of the tasks can
finish leading to a bunch of errors returned by the task group.

cognifloyd committed:
nfpm.native_libs.scripts: Customize User-Agent header

cognifloyd committed:
nfpm.native_libs.scripts: Move TODO about search API errors

cognifloyd committed:
nfpm.native_libs.scripts: Add from_best_so_files ld.so-like filtering

This only adds package dependencies without any version constraints.
It might be possible to add a version constraint based on the so_version.
But, this is good enough for now.

cognifloyd committed:
nfpm.native_libs: rename package scripts->deb

Next commit will move rules into that package

cognifloyd committed:
nfpm.native_libs: move search_for_sonames rule into deb package

cognifloyd committed:
nfpm.native_libs: register rules after moving to deb package

cognifloyd committed:
nfpm.native_libs: regenerate user_reqs.lock to include new deps

__________________________________________________________________
Lockfile diff: 3rdparty/python/user_reqs.lock [python-default]
__________________________________________________________________
==                      Added dependencies                      ==
‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
  aiohappyeyeballs               2.6.1
  aiohttp                        3.12.15
  aiohttp-retry                  2.9.1
  aiosignal                      1.4.0
  attrs                          25.4.0
  frozenlist                     1.8.0
  multidict                      6.7.0
  propcache                      0.4.1
  yarl                           1.22.0

cognifloyd committed:
nfpm.native_libs: Include native_libs.deb in docs/notes

cognifloyd committed:
nfpm.native_libs.deb: add rules to prevent py deps on :scripts

cognifloyd committed:
Revert "nfpm.native_libs.deb: add rules to prevent py deps on :scripts"

This reverts commit 44907754618524228661a6359b21ad4475ca74cd.

cognifloyd committed:
nfpm.native_libs.deb: add usage comments

cognifloyd commented on src/python/pants/backend/nfpm/native_libs/deb/search_for_sonames.py:
```diff
@@ -0,0 +1,209 @@
+            headers={aiohttp.hdrs.USER_AGENT: user_agent},
+            # version=aiohttp.HttpVersion11,  # aiohttp does not support HTTP/2 (waiting for contribution)
+            # timeout=aiohttp.ClientTimeout(total=5 * 60, sock_connect=30),
+        ) as client,
```
The script accepts a `--user-agent-suffix` so that we can make sure debian knows that this traffic is coming from pants. Hopefully if this causes any issues, they can reach out and ask for a different approach. Basically, my philosophy is to make sure I am blame-able (git blame is beautiful!) when I change things, to speed up the process of telling me about [inadvertent] issues so I can fix them ASAP.

Also, I have a `timeout` stubbed in here. The `timeout` arg (and `headers`) gets passed to an `aiohttp.ClientSession`. The commented value represents the default timeout (5 minutes with sock_connect timeout of 30 seconds). Should I just remove this comment, or do something else with it?

cognifloyd commented on src/python/pants/backend/nfpm/native_libs/deb/rules.py:
```diff
@@ -0,0 +1,225 @@
+    "/usr/lib/",
+    # Anything else has to be added manually to dependencies.
+    # These rules cannot use symbols or shlibs metadata to inform package selection.
+)
```
This is a simplified heuristic, compared to the actual logic used by `/lib/ld-linux.so` and similar dynamic ELF loaders.

I commented out the `/usr/local` variants because packages available via the debian/ubuntu package search APIs are unlikely to install anything in this sysadmin-managed directory.

cognifloyd reviewed (commented):
This PR was mostly extracted from #22861, which is far too big to ask anyone to review it. But, if you want more context about how these rules get used, you can look in #22861.

You can also look at #22905 the draft PR for the next (and last) part of the PR series to see how these rules tie into package dependency injection rules.

cognifloyd committed:
Merge branch 'main' into cognifloyd/nfpm-native_libs-deb-scripts

benjyw commented on src/python/pants/backend/nfpm/native_libs/deb/search_for_sonames.py:
```diff
@@ -0,0 +1,209 @@
+Rule code should run this script as a subprocess in a sandboxed venv. Generally, that venv should
+be a subset of the pants venv (using the version of packages that pants depends on).
+
+WARNING: This script relies on aiohttp and asyncio, so their are probably incompatibilities
```
```suggestion
WARNING: This script relies on aiohttp and asyncio, so there are probably incompatibilities
```

benjyw reviewed (commented):
Looks good overall!  The script setup is a little gnarly, but that is likely preferable to attempting to run this in-process.

cognifloyd committed:
nfpm.native_libs.deb: drop /usr/local comments

Though ld.so looks in /usr/local, packages from system package
repos are not supposed to install anything in /usr/local, so
searching for that does not make sense. Keeping it commented
also adds no value, so drop the commented /usr/local entries.

cognifloyd commented on src/python/pants/backend/nfpm/native_libs/deb/rules.py:
```diff
@@ -0,0 +1,225 @@
+        "aiohttp",
+        "aiohttp-retry",
+        "beautifulsoup4",
+    }
```
> This isn't REAPI friendly, for the reason you state (the python interpreter is an external dependency).

I was afraid of that. I wonder if we can use the PBS provider backend without forcing it to be enabled for all python rules?

> You might want to set the `process_execution_environment` arg to `execute_process()` to force it to run locally?

I hate to lose the potential benefits of REAPI. But, making that explicit would probably be wise if I stick with this approach.

cognifloyd replied:
I guess I really just need to add yet another lockfile for this. Avoiding the complexity of lockfiles (or of exposing too many knobs) just moved the complexity elsewhere.

cognifloyd committed:
nfpm.native_libs.deb: split tests into separate files

cognifloyd commented on src/python/pants/backend/nfpm/native_libs/deb/rules.py:
```diff
@@ -0,0 +1,225 @@
+        # A 4xx or 5xx error means we gave up retrying because the server is unavailable.
+        # TODO: Should this raise an error instead of just a warning?
+        logger.warning(result.stderr.decode("utf-8"))
+        packages = {}
```
I want to add a test that covers the error case, so I'm marking this as draft while I figure that out.

cognifloyd committed:
Merge branch 'main' into cognifloyd/nfpm-native_libs-deb-scripts

cognifloyd committed:
typo fix

cognifloyd committed:
nfpm.native_libs.deb: refactor script error handling

The API returns success (200) when returning an empty result set.
A not-found result is not a failure, as it means that the dep has to be
provided by a custom package; it just means the dep is not in one of the
distro-provided packages. So, now the script also returns success (0)
when an empty result-set is returned.

cognifloyd committed:
nfpm.native_libs.deb: new deb subsystem for configurable search URLs

cognifloyd committed:
Merge branch 'main' into cognifloyd/nfpm-native_libs-deb-scripts

cognifloyd committed:
nfpm.native_libs.deb: protect against "Client Challenge" response

cognifloyd committed:
nfpm.native_libs.deb: drop unused dict (moved to subsystem)

cognifloyd committed:
nfpm.native_libs.deb: comment out tests w/ blocked API requests

It seems debian added some kind of policy that requires a javascript-enabled
browser to get search results. Ubuntu has not done the same. Effectively, that
means that this script can only support ubuntu.

Also fix add <title> to the HTML parsing unit test since that is expected now.

cognifloyd committed:
Merge branch 'main' into cognifloyd/nfpm-native_libs-deb-scripts

cognifloyd commented:
I'm really sad about this PR. I've spent a lot of time trying to refactor this into something that is clean enough to incorporate into pants, but it's just not there. And now packages.debian.org is blocking the search requests making this less generally useful.

An alternative would be to use the pkgs.org API, but that looks like it requires an API key and possibly a paid "premium" membership. I don't see any other host-agnostic ways to do this.

It would be interesting to ask the debian webmasters about how to respectfully use the API, but I just can't justify spending more time on this. If anyone else is interested in this, please ask the debian webmasters about this (I've seen mention of some emails or mailing lists).

sureshjoshi commented:
Thanks for the contribution. We've just branched for 2.32.x, so merging this pull request now will come out in 2.33.x, please move the release notes updates to docs/notes/2.33.x.md if that's appropriate.
